src/pyhf/infer/intervals/upper_limits.py

```diff
@@ -0,0 +1,282 @@
+"""Interval estimation"""
+from warnings import warn
+
+import numpy as np
+from scipy.optimize import toms748
+
+from pyhf import get_backend
+from pyhf.infer import hypotest
+
+__all__ = ["upper_limit", "linear_grid_scan", "toms748_scan"]
+
+
+def __dir__():
+    return __all__
+
+
+def _interp(x, xp, fp):
+    tb, _ = get_backend()
+    return tb.astensor(np.interp(x, xp.tolist(), fp.tolist()))
+
+
+def toms748_scan(
+    data,
+    model,
+    bounds_low,
+    bounds_up,
+    level=0.05,
+    atol=2e-12,
+    rtol=None,
+    from_upper_limit_fn=False,
+    **hypotest_kwargs,
+):
+    """
+    Calculate an upper limit interval ``(0, poi_up)`` for a single
+    Parameter of Interest (POI) using an automatic scan through
+    POI-space, using the :func:`~scipy.optimize.toms748` algorithm.
+
+    Example:
+        >>> import numpy as np
+        >>> import pyhf
+        >>> pyhf.set_backend("numpy")
+        >>> model = pyhf.simplemodels.uncorrelated_background(
+        ...     signal=[12.0, 11.0], bkg=[50.0, 52.0], bkg_uncertainty=[3.0, 7.0]
+        ... )
+        >>> observations = [51, 48]
+        >>> data = pyhf.tensorlib.astensor(observations + model.config.auxdata)
+        >>> obs_limit, exp_limits = pyhf.infer.intervals.upper_limits.toms748_scan(
+        ...     data, model, 0., 5., rtol=0.01
+        ... )
+        >>> obs_limit
+        array(1.01156939)
+        >>> exp_limits
+        [array(0.5600747), array(0.75702605), array(1.06234693), array(1.50116923), array(2.05078912)]
+
+    Args:
+        data (:obj:`tensor`): The observed data.
+        model (~pyhf.pdf.Model): The statistical model adhering to the schema ``model.json``.
+        bounds_low (:obj:`float`): Lower boundary of search interval.
+        bounds_up (:obj:`float`): Upper boundary of search interval.
+        level (:obj:`float`): The threshold value to evaluate the interpolated results at.
+                              Defaults to ``0.05``.
+        atol (:obj:`float`): Absolute tolerance.
+                             Defaults to ``1e-12``.
+                             The iteration will end when the result is within absolute
+                             *or* relative tolerance of the true limit.
+        rtol (:obj:`float`): Relative tolerance.
+                             For optimal performance this argument should be set
+                             to the highest acceptable relative tolerance, though
+                             it will default to ``1e-15`` if not set.
+        hypotest_kwargs (:obj:`string`): Kwargs for the calls to
+         :class:`~pyhf.infer.hypotest` to configure the fits.
+
+    Returns:
+        Tuple of Tensors:
+
+            - Tensor: The observed upper limit on the POI.
+            - Tensor: The expected upper limits on the POI.
+
+    .. versionadded:: 0.7.0
```

Comment:
I put the `.. versionadded` at the bottom to match how the CPython docs use it. c.f. https://docs.python.org/3/library/shutil.html#shutil.copystat